File: backend/src/app/schemas/evaluations.py
"""评测相关 Schemas。"""

from __future__ import annotations

import uuid
from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class EvaluationStatus(str, Enum):
    QUEUED = "queued"
    RUNNING = "running"
    SUCCEEDED = "succeeded"
    FAILED = "failed"
    CANCELED = "canceled"


class EvaluationRunCreateRequest(BaseModel):
    """创建评测任务请求。"""

    dataset: dict[str, Any] = Field(..., description="问题集与评分规则")
    config: dict[str, Any] = Field(..., description="范围选择、模式等配置")


class EvaluationRunRead(BaseModel):
    """评测任务响应。"""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    status: EvaluationStatus
    summary: dict[str, Any] | None = None
    error_message: str | None = None
    created_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None


class EvaluationCaseResult(BaseModel):
    """单题评测结果。"""

    question_id: str
    question: str
    single_agent_run_id: uuid.UUID | None = None
    multi_agent_run_id: uuid.UUID | None = None
    single_agent_answer: str | None = None
    multi_agent_answer: str | None = None
    single_agent_metrics: dict[str, Any] | None = None
    multi_agent_metrics: dict[str, Any] | None = None
    reference_answer: str | None = None


class EvaluationResults(BaseModel):
    """评测结果汇总。"""

    eval_run_id: uuid.UUID
    status: EvaluationStatus
    summary: dict[str, Any] | None = None
    case_results: list[EvaluationCaseResult] = Field(default_factory=list)

File: frontend/src/services/evaluations.ts
/**
 * 评测 API 封装
 */

import { apiFetch } from './http';

export type EvaluationStatus = 'queued' | 'running' | 'succeeded' | 'failed' | 'canceled';

export interface EvaluationRunCreateRequest {
  dataset: {
    questions: Array<{
      id: string;
      question: string;
      reference_answer?: string;
      scoring_criteria?: Record<string, number>;
      tags?: string[];
    }>;
  };
  config: {
    selected_kb_ids: string[];
    allow_external?: boolean;
    timeout_per_question_ms?: number;
  };
}

export interface EvaluationRun {
  id: string;
  status: EvaluationStatus;
  summary: {
    total_questions?: number;
    single_agent?: {
      avg_score: number;
      avg_latency: number;
    };
    multi_agent?: {
      avg_score: number;
      avg_latency: number;
    };
    case_results?: Array<{
      question_id: string;
      question: string;
      single_agent_run_id?: string;
      multi_agent_run_id?: string;
      single_agent_answer?: string;
      multi_agent_answer?: string;
      single_score?: number;
      multi_score?: number;
      reference_answer?: string;
    }>;
  } | null;
  error_message: string | null;
  created_at: string;
  started_at: string | null;
  finished_at: string | null;
}

export interface EvaluationResults {
  eval_run_id: string;
  status: EvaluationStatus;
  summary: EvaluationRun['summary'];
  case_results: NonNullable<EvaluationRun['summary']>['case_results'];
}

/**
 * 发起对比评测
 */
export async function createEvaluationRun(data: EvaluationRunCreateRequest): Promise<EvaluationRun> {
  return apiFetch<EvaluationRun>('/api/v1/evaluations/runs', {
    method: 'POST',
    body: JSON.stringify(data),
  });
}

/**
 * 获取评测状态
 */
export async function getEvaluationRun(evalRunId: string): Promise<EvaluationRun> {
  return apiFetch<EvaluationRun>(`/api/v1/evaluations/runs/${evalRunId}`);
}

/**
 * 获取评测结果
 */
export async function getEvaluationResults(evalRunId: string): Promise<EvaluationResults> {
  return apiFetch<EvaluationResults>(`/api/v1/evaluations/runs/${evalRunId}/results`);
}

/**
 * 取消评测任务
 */
export async function cancelEvaluationRun(evalRunId: string): Promise<EvaluationRun> {
  return apiFetch<EvaluationRun>(`/api/v1/evaluations/runs/${evalRunId}/cancel`, {
    method: 'POST',
  });
}
